File: services/tool-service/app/services/schema_service.py
import httpx

from app.config import Settings
from app.schemas import ColumnSchema, GetSchemaResponse, TableSchema


class SchemaNotConfiguredError(Exception):
    pass


class SchemaRequestError(Exception):
    def __init__(self, message: str, status_code: int) -> None:
        super().__init__(message)
        self.status_code = status_code


class SchemaService:
    """Read-only introspection of the real Supabase schema — tables, columns,
    and constraints (primary keys, foreign keys, unique, and check constraints).

    This is the only place in the whole system that holds the Supabase secret
    key. It exists purely so the AI agent can look up real table/column names
    and constraints before generating code — it is never used for runtime
    artifact data access, which stays anon-key-only via supabase-service.
    """

    def __init__(self, settings: Settings) -> None:
        self._base_url = settings.supabase_url.rstrip("/") if settings.supabase_url else None
        self._secret_key = settings.supabase_secret_key

    def describe(self) -> GetSchemaResponse:
        if not self._base_url or not self._secret_key:
            raise SchemaNotConfiguredError("SUPABASE_URL / SUPABASE_SECRET_KEY are not configured")

        try:
            response = httpx.get(
                f"{self._base_url}/rest/v1/",
                headers={
                    "apikey": self._secret_key,
                    "Authorization": f"Bearer {self._secret_key}",
                    "Accept": "application/openapi+json",
                },
                timeout=15.0,
            )
        except httpx.HTTPError as exc:
            raise SchemaRequestError(f"Failed to reach Supabase for schema introspection: {exc}", 502) from exc

        if response.is_error:
            raise SchemaRequestError(f"Supabase schema request failed (status {response.status_code})", response.status_code)

        spec = response.json()
        definitions = spec.get("definitions") or spec.get("components", {}).get("schemas") or {}
        constraints_by_table = self._fetch_constraints()

        tables = []
        for table, definition in definitions.items():
            properties = definition.get("properties") or {}
            required = set(definition.get("required") or [])
            columns = [
                ColumnSchema(
                    name=name,
                    type=prop.get("format") or prop.get("type") or "unknown",
                    nullable=name not in required,
                )
                for name, prop in properties.items()
            ]
            tables.append(
                TableSchema(table=table, columns=columns, constraints=constraints_by_table.get(table, []))
            )

        return GetSchemaResponse(tables=tables)

    def _fetch_constraints(self) -> dict[str, list[str]]:
        """Constraints (primary keys, foreign keys, unique, and check) never
        appear in PostgREST's schema/OpenAPI document — only column
        name/type/nullable do. This calls a small SECURITY DEFINER RPC (see
        services/supabase-service/sql/003_create_table_constraints_rpc.sql),
        which is a thin wrapper around:

            SELECT conname AS constraint_name, contype AS constraint_type,
                   pg_get_constraintdef(c.oid) AS definition
            FROM pg_constraint c
            JOIN pg_class t ON c.conrelid = t.oid
            WHERE t.relname = '<table name>';

        run across every table in one call instead of one table at a time. If
        that migration hasn't been run yet, this degrades to "no constraints
        known" rather than failing the whole schema lookup.
        """
        try:
            response = httpx.post(
                f"{self._base_url}/rest/v1/rpc/get_table_constraints",
                headers={
                    "apikey": self._secret_key,
                    "Authorization": f"Bearer {self._secret_key}",
                },
                timeout=15.0,
            )
        except httpx.HTTPError:
            return {}

        if response.is_error:
            return {}

        by_table: dict[str, list[str]] = {}
        for entry in response.json() or []:
            by_table.setdefault(entry["table_name"], []).append(
                f"{entry['constraint_name']}: {entry['definition']}"
            )
        return by_table

File: services/db-agent-service/src/services/schema-service.ts
import type { AppConfig } from '../config.js';
import type { SupabaseQueryClient } from './supabase-query-client.js';

export interface ColumnRow {
  table_name: string;
  column_name: string;
  data_type: string;
  udt_name: string;
  is_nullable: boolean;
  column_default: string | null;
  ordinal_position: number;
}

export interface ConstraintRow {
  table_name: string;
  constraint_name: string;
  constraint_type: string;
  definition: string;
}

export interface EnumRow {
  enum_name: string;
  enum_value: string;
}

/** Raw, still-structured schema data for a single table — what form-spec-builder.ts needs to
 *  build a FormSpec. A subset of the same rows describe()'s cache already holds; see getTableInfo. */
export interface TableSchemaInfo {
  columns: ColumnRow[];
  constraints: ConstraintRow[];
  enums: EnumRow[];
}

const FALLBACK_DESCRIPTION =
  'Schema lookup failed — no table/column names are currently known. Ask the user what data ' +
  'they mean (table/column names) before calling query_table or write_table, and treat any ' +
  'guesses as unverified.';

/**
 * Builds the "known tables and columns" section of the system prompt by reading the live
 * Supabase schema, instead of a hand-maintained copy that can silently drift from what's
 * actually deployed (see git history: schema-context.ts's old static listing was written from
 * packages/sql/01_create_tables.sql, an aspirational schema that turned out not to match this
 * project's real Supabase instance at all — every query failed until that was caught by hand).
 *
 * Reads three read-only Postgres catalog RPCs — get_schema_columns, get_table_constraints,
 * get_enum_values (services/supabase-service/sql/00[3-5]_*.sql) — through supabase-service's
 * generic /rpc/:function proxy, under the caller's own JWT. All three are SECURITY DEFINER and
 * granted to `anon`/`authenticated`, so — deliberately, matching this service's "holds no
 * Supabase key at all" design — no secret key is needed anywhere in this path, unlike opencode's
 * get_schema tool which reads the same kind of information but at artifact-generation time via a
 * secret key it's allowed to hold because it's never in a runtime request path.
 *
 * Schema metadata is identical for every caller (these RPCs read Postgres's own catalogs, not
 * RLS-governed application tables), so the result is cached process-wide for AppConfig.schemaCacheTtlMs
 * rather than per-user or per-request — short enough to notice a real schema change on its own
 * within a few minutes, long enough that an ordinary multi-turn chat doesn't re-fetch it every turn.
 */
interface Cache {
  description: string;
  columns: ColumnRow[];
  constraints: ConstraintRow[];
  enums: EnumRow[];
  expiresAt: number;
}

export class SchemaService {
  private cached: Cache | null = null;

  constructor(
    private readonly supabaseQuery: SupabaseQueryClient,
    private readonly config: AppConfig,
  ) {}

  async describe(jwt: string): Promise<string> {
    const cache = await this.ensureCache(jwt);
    return cache?.description ?? FALLBACK_DESCRIPTION;
  }

  /** Structured schema data for a single table — what form-spec-builder.ts needs to build a
   *  FormSpec. Shares the exact same cache/fetch path describe() uses, so calling this after (or
   *  before) describe() in the same turn doesn't trigger a second round of RPC calls. */
  async getTableInfo(jwt: string, table: string): Promise<TableSchemaInfo> {
    const cache = await this.ensureCache(jwt);
    if (!cache) {
      return { columns: [], constraints: [], enums: [] };
    }
    return {
      columns: cache.columns.filter((c) => c.table_name === table),
      constraints: cache.constraints.filter((c) => c.table_name === table),
      enums: cache.enums,
    };
  }

  private async ensureCache(jwt: string): Promise<Cache | null> {
    if (this.cached && this.cached.expiresAt > Date.now()) {
      return this.cached;
    }

    let columns: ColumnRow[];
    try {
      columns = await this.supabaseQuery.callRpc<ColumnRow[]>(jwt, 'get_schema_columns');
    } catch (err) {
      console.error('SchemaService.describe: get_schema_columns failed, serving fallback:', err);
      // Serve a stale-but-real cache over the generic fallback if we have one.
      return this.cached;
    }

    // Constraints and enum values are enrichment, not the core listing — degrade quietly (empty)
    // if either RPC migration hasn't been applied yet, rather than failing the whole lookup,
    // mirroring get_schema.ts's own fallback behavior for the same two RPCs.
    const [constraints, enums] = await Promise.all([
      this.supabaseQuery.callRpc<ConstraintRow[]>(jwt, 'get_table_constraints').catch((err) => {
        console.warn('SchemaService.describe: get_table_constraints unavailable:', err);
        return [] as ConstraintRow[];
      }),
      this.supabaseQuery.callRpc<EnumRow[]>(jwt, 'get_enum_values').catch((err) => {
        console.warn('SchemaService.describe: get_enum_values unavailable:', err);
        return [] as EnumRow[];
      }),
    ]);

    const description = formatSchema(columns, constraints, enums);
    this.cached = { description, columns, constraints, enums, expiresAt: Date.now() + this.config.schemaCacheTtlMs };
    console.log(`SchemaService.describe: refreshed live schema (${columns.length} column row(s) across the public schema)`);
    return this.cached;
  }
}

function formatSchema(columns: ColumnRow[], constraints: ConstraintRow[], enums: EnumRow[]): string {
  if (columns.length === 0) {
    return 'No tables are currently exposed in the public schema.';
  }

  const constraintsByTable = groupBy(constraints, (c) => c.table_name);
  const enumValuesByType = new Map<string, string[]>();
  for (const row of enums) {
    const list = enumValuesByType.get(row.enum_name) ?? [];
    list.push(row.enum_value);
    enumValuesByType.set(row.enum_name, list);
  }
  const columnsByTable = groupBy(columns, (c) => c.table_name);

  const lines: string[] = [];
  for (const [table, cols] of columnsByTable) {
    const columnText = cols
      .map((col) => {
        // information_schema reports an enum column's data_type as the unhelpful literal
        // "USER-DEFINED" — udt_name is the actual type name (e.g. "user_role"), both for display
        // and for matching against get_enum_values' enum_name.
        const displayType = col.data_type === 'USER-DEFINED' ? col.udt_name : col.data_type;
        const enumValues = enumValuesByType.get(col.udt_name);
        const enumSuffix = enumValues?.length ? `, allowed: ${enumValues.join(', ')}` : '';
        return `${col.column_name} (${displayType}${col.is_nullable ? ', nullable' : ''}${enumSuffix})`;
      })
      .join(', ');
    lines.push(`- ${table}: ${columnText}`);
    for (const constraint of constraintsByTable.get(table) ?? []) {
      lines.push(`    CONSTRAINT — ${constraint.constraint_name}: ${constraint.definition}`);
    }
  }

  return (
    'Known tables and columns (CONSTRAINT lines are PRIMARY KEY / FOREIGN KEY / UNIQUE / CHECK ' +
    "constraints, plus enum-typed columns' allowed values — any row you write must satisfy every " +
    'listed constraint for that table):\n' +
    lines.join('\n')
  );
}

function groupBy<T>(rows: T[], key: (row: T) => string): Map<string, T[]> {
  const map = new Map<string, T[]>();
  for (const row of rows) {
    const k = key(row);
    const list = map.get(k) ?? [];
    list.push(row);
    map.set(k, list);
  }
  return map;
}
